### ISSUE TITLE
Output app list doesn't display icons of Flatpak applications

### ISSUE THREAD

**Issue by vchernin** [wwmm/easyeffects#1047]
In the output section, while audio can be manipulated for Flatpak apps, the icon for the app doesn't seem to show. This issue is specifically about other apps being a Flatpak, which affects both the Arch package and the EasyEffects Flatpak. 

Arch package with Chromium Flatpak installed:
![image](https://user-images.githubusercontent.com/38842733/127086556-63184a79-3bfd-4e1d-8192-ac72b8328be0.png)

Flatpak Package with Chromium Flatpak:
![image](https://user-images.githubusercontent.com/38842733/127086647-ea7a10d5-8173-40d6-8be1-0eb6c7a5ea04.png)

Tested on Arch with KDE, also occurs on GNOME.

**Comment by wwmm** [comment]
This is probably a gtk issue when running in flatpak. After getting the app icon name from PipeWire we just set this name in the image widget. gtk is the one actually finding the icon and loading it. Flatpak sandbox must be breaking whatever gtk is doing to load these icons.

**Comment by wwmm** [comment]
For example under normal circunstances Chromium icon can be found in a path similar to `/usr/share/icons/hicolor/256x256/apps/`. gtk finds icons in standard places on its own. But the flatpak package probably has them somewhere else. If gtk is supposed to be able to find them in the flatpak sandbox I do not know.

**Comment by vchernin** [comment]
The equivalent directory for Flatpaks are here. Note you shouldn't need to look inside the sandbox.

`/var/lib/flatpak/exports/share/icons/hicolor` (system wide installed Flatpaks) and `~/.local/share/flatpak/exports/share/icons/hicolor` (per user installed Flatpaks).

In each directory you will find a bunch of directories something like:
```
128x128  24x24    32x32  512x512  96x96             index.theme
16x16    256x256  48x48  64x64    icon-theme.cache  scalable
```
And similar to the traditional way each icon size has an `apps` directory within it, and that's where the icons are.
So for 256x256 icons:
`/var/lib/flatpak/exports/share/icons/hicolor/256x256/apps` and `~/.local/share/flatpak/exports/share/icons/hicolor/256x256/apps`

Since these are accesible without looking into the sandbox, it seems the issue is gtk just isn't aware where to look?

Further reference:
To install a Flatpak system wide: `flatpak install com.github.wwmm.easyeffects` 
To install a Flatpak per user: `flatpak install --user com.github.wwmm.easyeffects`

**Comment by wwmm** [comment]
>it seems the issue is gtk just isn't aware where to look?

It seems that way. Did this happen in the older releases? I set the icon names now in the same way as I did in gtk3. If only now that we are using gtk4 this problem is happening it is a toolkit regression.

**Comment by vchernin** [comment]
From this older PulseEffects Flatpak issue it seems like the "Flatpak Pulse/EasyEffects + Flatpak other app" case isn't a new issue: https://github.com/flathub/com.github.wwmm.pulseeffects/issues/4

For the "traditional package Pulse/EasyEffects + Flatpak other app" case I'll test out the PulseEffects Fedora package (version 5.0.4).

**Comment by wwmm** [comment]
I wonder if it will be necessary to call `gtk_icon_theme_add_search_path` manually https://developer.gnome.org/gtk4/stable/GtkIconTheme.html#gtk-icon-theme-add-search-path. But it still feels like something that gtk should be able to do on its own.

**Comment by vchernin** [comment]
This issue still occurs with the "traditional package PulseEffects (Fedora RPM 5.0.4 GTK 3) + Flatpak other app" case:
![image](https://user-images.githubusercontent.com/38842733/127258597-24c19a54-0ed8-4b53-9dc3-c87da3cf7a2a.png)

I will try to find other apps that do the same thing, surely someone else has run into this...

Indeed `gtk_icon_theme_add_search_path` sounds like it might help.

**Comment by vchernin** [comment]
> I wonder if it will be necessary to call `gtk_icon_theme_add_search_path` manually https://developer.gnome.org/gtk4/stable/GtkIconTheme.html#gtk-icon-theme-add-search-path. But it still feels like something that gtk should be able to do on its own.

FYI this link is now https://developer-old.gnome.org/gtk4/stable/GtkIconTheme.html#gtk-icon-theme-add-search-path. The developer site was reworked, but from what I can tell the API references haven't been moved yet.

I did also find this (new I think) site: https://developer.gnome.org/documentation/tutorials/themed-icons.html?highlight=gtkicontheme

Still I don't know why this isn't done automatically...

**Comment by vchernin** [comment]
Ah I finally thought of a decent reference: Flatseal, it's not a GTK4 app yet, and written in JS it seems, but hopefully it will help. I'm not familar with how much things will differ between languages.

https://github.com/tchx84/Flatseal/blob/fdb04dfc9d6ae70954bb4c1e9af77c819b619667/src/widgets/applicationRow.js#L24

It is able to display app icons within itself just fine.

I built flatseal with GTK Debug and inspected the icon, is there anything useful here?

![image](https://user-images.githubusercontent.com/38842733/130331912-58fe7c82-f12f-45e0-b288-5db6da365eb9.png)

In this case com.belmoussaoui.Decoder is the other app name (I just picked it at random).

**Comment by wwmm** [comment]
>is there anything useful here?

Nothing out of the ordinary.  Looking at this app code it seems to me that they are searching for possible paths in the system and passing them to gtk https://github.com/tchx84/Flatseal/blob/7988e1813b31d17e5227523a9b3ae3bdbf77b1f2/src/models/applications.js#L324. I can see this working. It just seems wrong that gtk or flatpak does not have a standard solution to this situation.

**Comment by vchernin** [comment]
> I can see this working. It just seems wrong that gtk or flatpak does not have a standard solution to this situation.

Yeah that appears to be the case unfortunately. Maybe later I can make a GTK bug report, perhaps there's a way of improving this.

**Comment by Digitalone1** [comment]
> I wonder if it will be necessary to call `gtk_icon_theme_add_search_path` manually https://developer.gnome.org/gtk4/stable/GtkIconTheme.html#gtk-icon-theme-add-search-path. But it still feels like something that gtk should be able to do on its own.

Not exactly. AFAIK, this is the standard followed by GTK and most of applications: https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html#directory_layout 

There's not mention about `/var/lib/flatpak`, so I think GTK won't ever be able to to this. It's Flatpak not following the standard. 

> FYI this link is now https://developer-old.gnome.org/gtk4/stable/GtkIconTheme.html#gtk-icon-theme-add-search-path. 

We use gtkmm, so the method is this: https://developer-old.gnome.org/gtkmm/stable/classGtk_1_1IconTheme.html#a06fbe6e5556cc1fc36c73394c4694ae8

I think this can be easily resolved installing a good icon theme different than Adwaita, anyway I'm curious to see what happens with that method, so I could add it in the next days and you can test the master branch on your system. But that would be only a workaround because this issue should be reported to Flatpak claiming that icons should be installed under `/usr/share/pixmaps` or `/usr/share/icons`.

**Comment by Digitalone1** [comment]
@wwmm In the next days, maybe tomorrow, I will make a big merge request with lots of changes to the code in order to improve it.

Then I'll try to implement some fix for the know issues. About this specific one, since it seems causes by Flatpak not installing the icons in the right place, I was wondering hot to detect if the application running is shipped by Flatpak.

If there's a way to do it, the fix of adding new path for searching icons would be applied only to Flatpak builds, not to the rest that do not need it.

**Comment by vchernin** [comment]
> If there's a way to do it, the fix of adding new path for searching icons would be applied only to Flatpak builds, not to the rest that do not need it.

The issue affects non-Flatpak builds of EasyEffects as well. The issue is that EasyEffects can't find the icons of _other_ apps that are installed with Flatpak.

For instance, [up here](https://github.com/wwmm/easyeffects/issues/1047#issuecomment-887983626) I tried the old Fedora RPM PulseEffects build. It can't find the Chromium Flatpak app icon just like the Flatpak version of EasyEffects can't. non-Flatpak and Flatpak builds of EasyEffects behave the same as far as I can tell.

It makes sense to add support for the Flatpak icon directories in all builds I think.

**Comment by wwmm** [comment]
>@wwmm In the next days, maybe tomorrow, I will make a big merge request with lots of changes to the code in order to improve it.

Ok.

>If there's a way to do it, the fix of adding new path for searching icons would be applied only to Flatpak builds, not to the rest that do not need it.

Any system without the Adwaita theme or without a way to find it (Flatpak) can have missing icons in our interface. The problem is that not all of the Adwaita icons have a replacement on other icon themes. The only way I see this being solved is embedding our icons in our executables like it used to be done in the past with the few custom icons the older releases had.

**Comment by vchernin** [comment]
> Any system without the Adwaita theme or without a way to find it (Flatpak) can have missing icons in our interface. The problem is that not all of the Adwaita icons have a replacement on other icon themes. The only way I see this being solved is embedding our icons in our executables like it used to be done in the past with the few custom icons the older releases had.

This is notably the cause of https://github.com/wwmm/easyeffects/issues/1050.

I will note though, while embedding the icon in the executable might make sense for https://github.com/wwmm/easyeffects/issues/1050, I think for https://github.com/wwmm/easyeffects/issues/1047 EasyEffects needs to be taught where to find the icons.  Fixing both sisues by embedding the icons sounds like a maintenance burden to me.

**Comment by wwmm** [comment]
>Fixing both issues by embedding the icons sounds like a maintenance burden to me.

Although I am not a big fan of this solution it will solve both problems at the same time.  And it would be a burden when doing it by the first time. In order to embed the icons we have to copy all the icons we use to our data folder. But after that we won't have to touch them anymore unless we want to change the icon. Meson will do the work.

Even if other icon themes have icons with the same name the icon design can be quite different. And as a result the user experience when navigating through our interface may not be the one intended. As far as I remember PulseEffects on KDE always had some "weird icons" in the window. It is better than having no icon at all. But it can still be bad.

**Comment by vchernin** [comment]
> Although I am not a big fan of this solution it will solve both problems at the same time. 

To be clear, I was referring to solving both #1050 and #1047 by embedding the icons, which in my eyes doesn't make sense. The problem for #1047 (this issue) is that we would effectively need to include every Flatpak app's icon that could possibly play audio. I am only referring to the external app icons that show up in the interface, [like Chromium that is supposed to be here](https://github.com/wwmm/easyeffects/issues/1047#issue-953449448). We have no control over what other apps do and it is hard to keep track of so many external apps and their icons, so there will inevitably be other app icons that we miss. It seems much easier to just add a [icon search path](https://github.com/wwmm/easyeffects/issues/1047#issuecomment-903161048) as mentioned above.

> Even if other icon themes have icons with the same name the icon design can be quite different. And as a result the user experience when navigating through our interface may not be the one we intended. As far as I remember PulseEffects on KDE always had some "weird icons" in the window. It is better than having no icon at all. But it can still be bad.

Yeah all this is very relevant for #1050.

**Comment by Digitalone1** [comment]
> The issue affects non-Flatpak builds of EasyEffects as well. The issue is that EasyEffects can't find the icons of _other_ apps that are installed with Flatpak.
>
> For instance, [up here](https://github.com/wwmm/easyeffects/issues/1047#issuecomment-887983626) I tried the old Fedora RPM PulseEffects build. It can't find the Chromium Flatpak app icon just like the Flatpak version of EasyEffects can't. non-Flatpak and Flatpak builds of EasyEffects behave the same as far as I can tell. [I also tested it here](https://github.com/wwmm/easyeffects/issues/1047#issue-953449448).
> 
> It makes sense to add support for the Flatpak icon directories in all builds I think.

OK, so it's better to add for all the flatpak path where those icons are installed.

**Comment by Digitalone1** [comment]
> Any system without the Adwaita theme or without a way to find it (Flatpak) can have missing icons in our interface. The problem is that not all of the Adwaita icons have a replacement on other icon themes. The only way I see this being solved is embedding our icons in our executables like it used to be done in the past with the few custom icons the older releases had.

I don't agree, this would break the style used by the icon theme. If an icon is not used by most popular icon themes (adwaita, breeze, papirus, kora), just change them with other ones. 

For app icons, the paths to add would be two, one system wide and the other under user locale. Should not be so hard.

**Comment by vchernin** [comment]
> OK, so it's better to add for all the flatpak path where those icons are installed.

Yes, you might want to see the [Flatseal reference](https://github.com/wwmm/easyeffects/issues/1047#issuecomment-903161048) for more info. 

However do note it can be confusing dealing with the directories with Flatpak. For the case when EasyEffects itself is in Flatpak, I think the Flatseal reference is enough.

However, I am not exactly sure if what Flatseal does is enough for non-Flatpak builds, since Flatseal is only available as a Flatpak and thus doesn't worry about if it's run in the host.

So when EasyEffects is not run in Flatpak on the host you might have to add directories like the ones mentioned [here](https://github.com/wwmm/easyeffects/issues/1047#issuecomment-887742273).

tl;dr the most complete fix for all builds might involve following Flatseal's example (or similar) **and** [adding other directories](https://github.com/wwmm/easyeffects/issues/1047#issuecomment-887742273) that _might_ only exist on the host.

> I don't agree, this would break the style used by the icon theme. If an icon is not used by most popular icon themes (adwaita, breeze, papirus, kora), just change them with other ones.

Do you mean to change the icon to something that has a valid icon in all the popular icon themes?

**Comment by wwmm** [comment]
>To be clear, I was referring to solving both #1050 and #1047 by embedding the icons, which in my eyes doesn't make sense. The problem for #1047 (this issue) is that we would effectively need to include every Flatpak app's icon that could possibly play audio.

Only 1047 would require including all icons. In that particular case I agree that embedding is not a solution. 1050 is another kind of issue. The icon not being shown is not from third party apps.

>I don't agree, this would break the style used by the icon theme. If an icon is not used by most popular icon themes (adwaita, breeze, papirus, kora), just change them with other ones.

That assumes that an icon theme will magically have the icon we need. This may or may not be the case. In the past I created an custom icon to represent an sine wave. Another one I needed at the time was an icon similar to "wave package" to represent an impulse response file. These are not the kind of icon that most icon themes if any will have. There are situations where a custom icon may be needed and in this case embedding is the better option whether we like or not the fact it will "break" the icon theme style. The only reason why I am not "breaking them" right now is because I decided not using those two icons anymore.

Just to be clear I don't like the idea of embedding icons in the executable and I will gladly avoid it if other alternatives are available. But there are situations where we can not expect third party icon themes to have a suitable replacement that will have the same technical meaning. It is not about just "the looks".  A few years ago when I saw the icons that were show in KDE for PulseEffects the thought that came to my mind was "this is horrible". The message those icons were giving to the user were not the ones I had in mind. I do not know how it is now but there is no point in respecting the local icon theme if the meaning of the icon is different.

**Comment by vchernin** [comment]
> Just to be clear I don't like the idea of embedding icons in the executable and I will gladly avoid it if other alternatives are available. But there are situations where we can not expect third party icon themes to have a suitable replacement that will have the same technical meaning. It is not about just "the looks".  A few years ago when I saw the icons that were show in KDE for PulseEffects the thought that came to my mind was "this is horrible". The message those icons were giving to the user were not the ones I had in mind. I do not know how it is now but there is no point in respecting the local icon theme if the meaning of the icon is different.

There's [some screenshots on KDE above](https://github.com/wwmm/easyeffects/issues/1047).

I do wonder, what do you think about not even supporting theming in general? [People even made a website about not theming apps](https://stopthemingmy.app/). Or is that just not a necessary or good idea here, and it's best to just try and fix edge cases where they're reported.

**Comment by wwmm** [comment]
>There's some screenshots on KDE above.

It is decent. I d not like the icon in the PipeWire section but in this case is just a matter of personal taste. It is not like the original icon has a super obvious meaning that was changed.

>I do wonder, what do you think about not even supporting theming in general? People even made a website about not theming apps. 

That solution seems too aggressive to me. I see it as that kind of last resort you only take when there is nothing else that can be done.

**Comment by Digitalone1** [comment]
@wwmm  Flatpak app icons apart, which I still think it's a Flatpak issue that should be resolved their side (but we can make a temporary workaround), I see there's only one problematic icon, the one on the plugin button. There should be an icon available on all major icon themes to replace it.

Setting a custom icon, at the risk of breaking icon theme style, it's not worth the effort considering that you also have to provide an alternative for dark mode and modify it on mode changes, which is not automatically related to the dark mode switch in EE settings (i.e. I have the switch disabled, but I still see the dark mode because I have set it at system level).  

These tasks are quite complicated and should be left to icon themes and GTK engine.

**Comment by wwmm** [comment]
>Flatpak app icons apart, which I still think it's a Flatpak issue that should be resolved their side

I totally agree. I think it is a little bizarre that Flatpak expects applications to search for its custom paths when the one that came after historically is it. `/usr/share/icons` is a standard path since forever. I have no idea about the technical challenges involved but in my opinion Flatpak is the one that should be handling things in a way that is compatible with what already existed before it was created. But we will probably have to do like other people and workaround the current situation =/

>Setting a custom icon, at the risk of breaking icon theme style, it's not worth the effort considering that you also have to provide an alternative for dark mode and modify it on mode changes

As surprising as it may seem the custom icons I used before automatically switched colors when a dark theme was used. The way to achieve this has changed in gtk4. So I do not know how it is done now. But in gtk3 it was quite simple. When doing a `-symbolic.svg` icon it was just a matter of setting very specific color in the svg. GT3 detected that color and understood that we wanted the icon to automatically switch its color when a dark mode was used.

**Comment by wwmm** [comment]
>GTK3 detected that color and understood that we wanted the icon to automatically switch its color when a dark mode was used.

Actually that is the reason why I use `-symbolic.svg` everywhere. It is not the icon theme that is magically doing the icon style switch. The icon is the same. The "color trick" is what makes gtk switch the mode on demand.

**Comment by vchernin** [comment]
So I had a long discussion on the Flatpak Matrix channel about this and got some very helpful information.

**tl;dr for implementation purposes: EasyEffects Flatpak should not show other app icons at all, and EasyEffects non Flatpak should look at these directories**

Longer explanation:

Flatpak does not really actually have a proper way to find other Flatpak apps’ icons. By default, Flatpak apps get access to `/usr/share/icons` (letting it find non Flatpak icons) but nothing else really. 

Again, there is no standard way of finding other Flatpak apps’s icons from within the Flatpak sandbox. Flatseal is a special case since it is really needs to be able to access the host and control things most Flatpak apps can’t. Flatseal gets extra file system permission just to access things like other Flatpak app icons, which EasyEffects Flatpak does not, and I would not like to change the file system access just for icons.

This is the case for Flatpak apps by default since it doesn’t want apps to be able to learn about other Flatpak apps. Letting Flatpak apps just look through icon directories of other Flatpaks lets an app find out all the names of the Flatpak apps you’ve installed.

There is currently no intention of adding a standard way of finding the other Flatpaks app icons due to the above reason.

So I think the cleanest solution is just to disable the icon feature when running in Flatpak. I think it’s best to not show them for non Flatpak apps as well for consistency. The easiest way and recommend way to check whether EasyEffects is being run in Flatpak by checking the existence of `/.flatpak-info`.

However for the non Flatpak packages (eg. Arch), since we can access everything anyway so there isn’t much of a problem. There you should be able to look at `XDG_DATA_DIRS` and find most Flatpak app icons, or just manually look at the directories listed in this comment. Technically there will be cases where users will install Flatpaks in non standard directories, but I don’t think it’s worth worrying about that.

**Comment by wwmm** [comment]
This is one of the moments where I philosophically disagree with how Flatpak is designed. The fact that the applications have to adapt to it does not happen with any of the standard package management systems. If for some weird reason the community decides in the future that something like Snap or Appimage is the way to go all the changes done in the application side specifically for Flatpak will be useless...

>The easiest and recommend way to check whether EasyEffects is being run in Flatpak by checking for the existence of /.flatpak-info.

I am not sure I understand. This is a file that exists only in the sandbox path?

>There you should be able to look at XDG_DATA_DIRS

At least on my system I do not see this environment variable in the output of `printenv`.

**Comment by wwmm** [comment]
>Either way if that's too confusing to figure out the directories listed in the comment above can be used, that's the standard expectation for Flatpak and most won't deviate from that.

It will be better doing this. It does not seem we can rely on this variable. I have other XDG variables defined. So I do not think glib has to be involved for this. XDG variables are used by other desktops besides gnome.

**Comment by Digitalone1** [comment]
The funny thing is that if you install a complete icon theme (which Adwaita is not), you can still see Flatpak apps icon in Flatpak EasyEffects because they are installed under `/usr/share/icons/`. This was a recommendation that I gave to @vchernin but he didn't try.

**Comment by wwmm** [comment]
>This was a recommendation that I gave to vchernin but he didn't try.

I do not know if this is something he can enforce from a package management point of view. Even if he bundles an icon theme in the flatpak package the theme will be restricted to the sandbox. It is probably also not possible to make the Flatpak package to somehow require the user to install an icon theme through its system standard package manager.

We can suggest the users to do that of course. But I do not see how a third party icon theme installation could be forced through the Flatpak package.

**Comment by wwmm** [comment]
Or could it be that icon themes installed through Flatpak do not put their files inside a sandbox?

**Comment by wwmm** [comment]
Maybe if the theme is installed to the same sandbox as EasyEffects this could work. But would it be really worth? IF a new applications is made and the icon theme is not updated with a new icon for it the problem will be there again. "Full icon themes" will work as long as someone decides to make a custom icon for a specific app. This may or may not happen depending on how popular an app is.

**Comment by Digitalone1** [comment]
I don't know how Flatpak works in detail, but as far as I know the icon theme is set at system level, so the application should know which is the icon theme even if it's running in a sandbox.

It could be also that Flatpak is giving their apps a specific environment variable to set a custom icon theme. But it should be installed outside the sandbox, otherwise no icon should be displayed at all.

**Comment by vchernin** [comment]
>  It is probably also not possible to make the Flatpak package to somehow require the user to install an icon theme through its system standard package manager.

Yup. As far as I know Flatpak has 0 ability to modify things like `/usr/share/` at all. 

> I don't know how Flatpak works in detail, but as far as I know the icon theme is set at system level, so the application should know which is the icon theme even if it's running in a sandbox.

For GTK themes, usually the way it works is you'd have to install the theme as a Flatpak. [But this is only for GTK themes really](https://blog.tingping.se/2017/05/11/flatpak-theming.html). For example if you're on Ubuntu you'd want to make sure the Yaru Flatpak theme is installed so Flatpak apps have the Yaru look.

I am not exactly sure how icon themes work with Flatpak or if they even exist. I only found this [doc page](https://docs.flatpak.org/en/latest/desktop-integration.html#other-notes-on-theming).

When I have time later I'll try playing around with Flatpak icon themes, and maybe I'll also finally try the suggestion to install an icon theme in `/usr/share/icons` and seeing what happens. I might have to ask in the Flatpak channel again. Hopefully I can give clearer info then. 

It might be Flatpak can't help us much with icon themes, and we'd either have to bundle in the executable, or tell users to install them via their system package manager, or just not show external app icons in Flatpak.

**Comment by vchernin** [comment]
Ok so indeed installing [Kora](https://github.com/bikass/kora) at `/usr/share/icons` is indeed a possible workaround. But I had to also change the default icon theme before it took any effect (in GNOME Tweaks). Notice this causes other icons to change.
![image](https://user-images.githubusercontent.com/38842733/131604038-4ff68bb5-1346-4641-95de-a80ec8ab23c4.png)

From poking a bit further it really doesn't seem like Flatpak icon themes are a thing. The user is expected to put whatever icon theme they like at `/usr/share/icons` as usual. But there is nothing we can easily we can integrate an icon theme in EasyEffects Flatpak, and there is basically nothing a Flatpak packager can do to conveniently steer people towards installing icon themes.

The ways forward for EasyEffects Flatpak that I see:

1. Embed something like Kora or another set of icons into the executable.
2. Leave instructions for users saying they could install a custom icon theme into `/usr/share/icons` and select that.
3. Don't show other app icons in Flatpak at all for consistency.
4. Try to show icons in Flatpak by default. When an icon can't be found don't show a missing icon picture.

I personally prefer 3 here given the circumstances. It seems too confusing from a user's perspective that some apps might show icons in EasyEffects and some won't. It will create noticeable difference between Flatpak and non-Flatpak builds, but I just don't see the other solutions being very ideal. But that's just my opinion.

What to do with the non-Flatpak builds seems a bit simpler, it just needs to be taught where Flatpak likes to put its icons from the host's perspective.

**Comment by wwmm** [comment]
Embedding a whole icon theme in the executable is very far from viable. When I was thinking about this I had in mind just the icons we use in the interface. Option 3 seems the better thing to do.

**Comment by Digitalone1** [comment]
There should be a way to know if the icon can be loaded from the system and we can show it on the correct result.

I don't like that kora users (or other complete icon themes like Papirus) can't view the icon only because they're using the Flatpak app. I'll try to find a workaround.

**Comment by Digitalone1** [comment]
I think I find the solution to this issue. At least it's working on my system, it has to be tested on others.

For what has been said, we don't have to care about Flatpak behaviour. It's responsibility of the user to install the proper icon them to show application icons correctly. Full icon themes like Papirus and kora can provide icons also for sandboxed applications in Flatpak.

So the only thing we can do it to detect if an icon name can be loaded, which means if the icon is present inside the icon theme set at system level at a specific moment.

Gkt provides an icon them class for this purpose. I hope it's working also outside my system.

@vchernin can you test after the merge?

**Comment by wwmm** [comment]
>Full icon themes like Papirus and kora can provide icons also for sandboxed applications in Flatpak.

If the app is popular enough for someone to decide to make a custom icon for it. Although I imagine that they will try to offer icons to as much apps as possible I doubt that they have custom icons for all the apps available on Linux.

**Comment by vchernin** [comment]
So this works as intended, thanks for fixing it!

I don't see any missing icons anymore when just using Adwaita. If select Kora as the icon theme then the Kora icons show up.

I did find if I open EasyEffects with Kora icon theme selected in GNOME Tweaks, but then select Adwaita, the missing icon image returns (when it should have shown nothing). But I think this is fine since a restart of EasyEffects fixes it, and very few people will be tweaking their icon theme while running apps and expecting everything to adjust correctly.

I think I'll close this as I doubt there is much else that can be done here.

**Comment by Digitalone1** [comment]
@vchernin the missing icon is shown because the name does not change and the check is not redone (the function that updates the app info ui is invoked many times, so I didn't want to make the check every time to save cpu cycles). If the player is restarted, the icon should be hidden looking up the new icon theme I think.

@wwmm can this be reopened to talk about the missing icons in other places of the interface?

I think it could be resolved choosing icon names present in the most used icon themes.

**Comment by wwmm** [comment]
>I think it could be resolved choosing icon names present in the most used icon themes.

It seems a good idea. If we are not going to embed icons in our executable we should at least choose icon names with a higher chance of being available everywhere.

**Comment by vchernin** [comment]
> >I think it could be resolved choosing icon names present in the most used icon themes.
> 
> 
> 
> It seems a good idea. If we are not going to embed icons in our executable we should at least choose icon names with a higher chance of being available everywhere.

Is there a good way of determining what icons names are present in what themes? E.g. the puzzle piece one in #1050.

You could always manually look them up and verify it works by hand but that seems like not an ideal method.

**Comment by Digitalone1** [comment]
Just search under `/usr/share/icons`

A good name would be inside at least Adwaita, Gnome, Breeze and Papirus. I'd say also Kora, but I think all the icons in Kora are already in Papirus.

**Comment by vchernin** [comment]
> Just search under `/usr/share/icons`
> 
> 
> 
> A good name would be inside at least Adwaita, Gnome, Breeze and Papirus. I'd say also Kora, but I think all the icons in Kora are already in Papirus.

I mean how do we make sure the icons we need are in the themes users will install. We can't feasibly manually install every single one of them out there to `usr/share/icons` and verify all of them work properly. 

But yes if the major ones are covered that is probably fine.

**Comment by Digitalone1** [comment]
For icons you have to choose conventional names. I don't know why the one of the puzzle is not showing. I'll take a look at it, but we have to go on symbolics which are the most popular.

**Comment by wwmm** [comment]
Some names come from the freedesktop standard https://specifications.freedesktop.org/icon-naming-spec/latest/ar01s04.html. As icon themes are supposed to follow this standard I think we should try to use only the names in this list. I wonder if a newer list somewhere else. This freedesktop page haven't been updated in many years.

**Comment by Digitalone1** [comment]
I'd say, like many does, just test on Gnome and Plasma ad default settings, if it's working for them both, it's okay.

**Comment by Digitalone1** [comment]
Unfortunately there's no puzzle icon with the same name for both Adwaita and Breeze. An option would be to change the icon to a "plus sign", available for both, but it's not really good.

At last I decided to add a workaround to detect if the icon is not showing and replace it with another available in Breeze. We are lucky this is happening in the same classes where the icon_theme object was defined.

**Comment by wwmm** [comment]
>Unfortunately there's no puzzle icon with the same name for both Adwaita and Breeze. An option would be to change the icon to a "plus sign", available for both, but it's not really good.

One possibility is taking a look at other icons. It does not have to a puzzle icon. As none of the standard icons have the meaning "plugins" we are totally free to use something else that looks decent in this position. The problem is that the puzzle icon is probably the closest to the idea of "plugging something somewhere".

**Comment by wwmm** [comment]
>The problem is that I discovered also the drag icon in the plugin stack is missing in Breeze.

Even this one? Well... I think that there is another icon that is also used for list item drags. Maybe this other icon has a replacement on other themes.

**Comment by Digitalone1** [comment]
Yes, I found one, `view-app-grid-symbolic`.

**Comment by Digitalone1** [comment]
I tried `view-app-grid-symbolic`, but it's too small compared to remove icon `user-trash-symbolic`. I made some tests and the better combination for both Adwaita and Breeze is `edit-delete-symbolic` and `view-app-grid-symbolic`.
